### PR TITLE
[4.0] Refactor Generator.Bind

### PR DESCRIPTION
**Pull request by Nihlus**
### Purpose of this PR
This PR heavily refactors Generator.Bind, effectively rewriting it completely. The new binder is faster, smarter, and more reliable.

In summary, the PR fixes the following issues or implements the following improvements:
* The binder now runs all generation tasks in parallel
* Generation times are down to ~70s for *all* targets, down from ~70s per target.
* Correctness is far more stringently checked. Invalid overrides will cause exceptions.
* The override and signature files now have XSD definitions.
* General structure and maintainability changes.

### Testing status

* Tested by diffing the API signatures of various profiles. There are differences, but they are either more correct than the old signatures, or simple naming changes.

### CODE REVIEW

**Comment by FreezyLemon** [comment]
Nice! Not done with the review yet, but wanted to ask something for now:
Seeing as you started using the JetBrains code annotations, I'd assume you also want to start using code inspection in CI in the near future? Or do you think StyleCop is "enough"?

By the way, it'd be nice if you could add the `Docs` directory to the "generated files and directories" list because that one alone has over 500 inspection results.

**Comment by Nihlus** [comment]
@FreezyLemon The JetBrains code inspections can't really be integrated into CI, so we'll have to live with StyleCop being our main check.

As for the docs directory, could you clarify what you mean?

**Comment by FreezyLemon** [comment]
Right, I didn't really explain that well...

R# has an option to ignore certain files and directories ("Generated Code"), and I thought you should ignore the folder with the docs since it's 
1. from an external source and
2. finds like 500-1000 inspection results which are mostly CSS, HTML and JS errors.

I've seen a way to use inspectcode in CI, I'll try and find the project that did it (and probably continue this in Discord)

**Comment by tzachshabtay** [comment]
Osu framework has inspectcode in their ci:https://github.com/ppy/osu-framework/blob/master/appveyor.yml

**Comment by Nihlus** [comment]
Is inspectcode available on Linux? Ought to be via Rider - out of scope for this PR, though.

**Comment by daerogami** [comment]
@Nihlus Unlikely related to your changes but I am getting the following on run:

'''System.IO.DirectoryNotFoundException
  HResult=0x80070003
  Message=Could not find a part of the path 'D:\repositories\opentk\src\Generator.Bind\bin\Debug\net461\src\Generator.Bind\Specifications\GL2\GL'.
  Source=mscorlib
  StackTrace:
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileSystemEnumerableIterator`1.CommonInit()
   at System.IO.FileSystemEnumerableIterator`1..ctor(String path, String originalUserPath, String searchPattern, SearchOption searchOption, SearchResultHandler`1 resultHandler, Boolean checkHost)
   at System.IO.Directory.GetFiles(String path, String searchPattern, SearchOption searchOption)
   at Bind.Generators.GL2.GL2GeneratorSettings..ctor() in D:\repositories\opentk\src\Generator.Bind\Generators\GL2\GL2GeneratorSettings.cs:line 50
   at Bind.Program.CreateGenerators() in D:\repositories\opentk\src\Generator.Bind\Program.cs:line 163
   at Bind.Program.<Main>d__7.MoveNext() in D:\repositories\opentk\src\Generator.Bind\Program.cs:line 76
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Bind.Program.<Main>(String[] args)
'''

Not sure if there is either some initial setup I am forgetting or if there is a bad path assumption.

**Comment by Nihlus** [comment]
@daerogami The default working directory is the location of the binary, while the binary defaults to assuming that it is being run in the root directory of the repository. You can change the working directory in the debugger to fix it :)

**Comment by Nihlus** [comment]
Very. I've been refactoring with other APIs in mind, such as EGL, Vulkan,
or OpenCL.

File: props/jetbrains.props [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<Project>
```
Comment (FreezyLemon) [inline props/jetbrains.props]: The other .props files use 2-space indentation. Would be nice if we could keep this consistent.

File: src/Generator.Bind/Program.cs [hunk 2]
```diff
@@ -79,38 +177,38 @@ private static void CreateGenerators()
                     {
                         case TargetAPI.GL2:
                         {
-                            Generators.Add(new GL2Generator());
+                            Generators.Add(new GL2GeneratorSettings());
                             break;
                         }
                         case TargetAPI.GL3:
                         case TargetAPI.GL4:
```
[the remark this answers is not in the document]
Comment (Nihlus) [reply src/Generator.Bind/Program.cs]: Yeah, that's true. Eventually, I'm hoping that GL3 and GL4 will actually generate different profiles.

[the remark this answers is not in the document]
Comment (FreezyLemon) [reply src/Generator.Bind/Program.cs]: Oh, okay. This is fine then, it wasn't a big deal in the first place

**Comment by DylanFPS** [comment]
@Nihlus Regarding the `inspectcode` command, it's not available on Linux, but seeing you're using AppVeyor (Windows-based), you will be able to use `inspectcode`

File: src/Generator.Bind/Program.cs [hunk 3]
```diff
@@ -41,35 +76,98 @@ private static int Main(string[] args)
...
             CreateGenerators();
 
-            foreach (var generator in Generators)
+            var stopwatch = Stopwatch.StartNew();
+            var generatorTasks = Generators.Select(GenerateBindingsAsync);
+            await Task.WhenAll(generatorTasks);
+            stopwatch.Stop();
+
+            Console.WriteLine();
+            Console.WriteLine("Bindings generated in {0} seconds.", stopwatch.Elapsed.Ticks / 10000000.0);
```
[the remark this answers is not in the document]
Comment (Nihlus) [reply src/Generator.Bind/Program.cs]: Good question :P

**Comment by varon** [comment]
@Nihlus  - I can commit some time to reviewing this. From your side, is this still a WIP or is it ready to be reviewed?

**Comment by Nihlus** [comment]
@varon Ready to be reviewed :)

**Comment by varon** [comment]
About to tackle this, but have a few questions:

- Does it generate bindings that are aligned with the previous ones?
- What testing has been done with the new bindings?
- Does it fulfill the requirements of the bindings generator?
- How extensible is it? (Vulkan? OpenAL?)
- What future extensions do you think we'd want to add?

**Comment by varon** [comment]
Requirements = actually works :).

Thanks for the fast answer. That all sounds good. I'll go ahead with the review.

**Comment by Nihlus** [comment]
Yes, it does actually work lol

**Pull request merged**